src/commons/util/file.rs

```diff
@@ -46,81 +55,95 @@ pub fn file_path(base_path: &Path, file_name: &str) -> PathBuf {
 }
 
 /// Saves a file, creating parent dirs as needed
-pub fn save(content: &[u8], full_path: &Path) -> Result<(), io::Error> {
+pub fn save(content: &[u8], full_path: &Path) -> Result<(), KrillIoError> {
     let mut f = create_file_with_path(full_path)?;
-    f.write_all(content)?;
+    f.write_all(content)
+        .map_err(|e| KrillIoError::new(format!("Could not write to: {}", full_path.to_string_lossy()), e))?;
 
     trace!("Saved file: {}", full_path.to_string_lossy());
     Ok(())
 }
 
 /// Saves an object to json - unwraps any json errors!
-pub fn save_json<O: Serialize>(object: &O, full_path: &Path) -> Result<(), io::Error> {
+pub fn save_json<O: Serialize>(object: &O, full_path: &Path) -> Result<(), KrillIoError> {
     let json = serde_json::to_string(object).unwrap();
     save(&Bytes::from(json), full_path)
 }
 
-/// Loads a files and deserializes as json for the expected type. Maps json
-/// errors to io::Error
-pub fn load_json<O: DeserializeOwned>(full_path: &Path) -> Result<O, io::Error> {
+/// Loads a files and deserializes as json for the expected type. Maps json errors to KrillIoError
+pub fn load_json<O: DeserializeOwned>(full_path: &Path) -> Result<O, KrillIoError> {
     let bytes = read(full_path)?;
-    serde_json::from_slice(&bytes).map_err(|_| io::Error::new(io::ErrorKind::Other, "could not deserialize json"))
+    serde_json::from_slice(&bytes).map_err(|_| {
+        KrillIoError::new(
+            format!("Could not load json for file: {}", full_path.to_string_lossy()),
+            io::Error::new(io::ErrorKind::Other, "could not deserialize json"),
+        )
+    })
 }
 
 /// Saves a file, creating parent dirs as needed
-pub fn save_in_dir(content: &Bytes, base_path: &Path, name: &str) -> Result<(), io::Error> {
+pub fn save_in_dir(content: &Bytes, base_path: &Path, name: &str) -> Result<(), KrillIoError> {
     let mut full_path = base_path.to_path_buf();
     full_path.push(name);
     save(content, &full_path)
 }
 
 /// Saves a file under a base directory, using the rsync uri to create
 /// sub-directories preserving the rsync authority and module in dir names.
-pub fn save_with_rsync_uri(content: &Bytes, base_path: &Path, uri: &uri::Rsync) -> Result<(), io::Error> {
+pub fn save_with_rsync_uri(content: &Bytes, base_path: &Path, uri: &uri::Rsync) -> Result<(), KrillIoError> {
     let path = path_with_rsync(base_path, uri);
     save(content, &path)
 }
 
 /// Reads a file to Bytes
-pub fn read(path: &Path) -> Result<Bytes, io::Error> {
-    let mut f = File::open(path).map_err(|_| Error::cannot_read(path))?;
+pub fn read(path: &Path) -> Result<Bytes, KrillIoError> {
+    let mut f =
+        File::open(path).map_err(|e| KrillIoError::new(format!("Could not read: '{}'", path.to_string_lossy()), e))?;
```

Comment:
Will `e` be sufficient to distinguish the failure to open from the failure to read, or should this message be different to that a few lines below?

Reply:
Changed first from 'read' to 'open'.